Added manual Saboteur cloaking

Mailaender opened the pull request:
Closes #10083.

Mailaender committed:
Make vehicles not demolishable.

Mailaender committed:
Add a charge drain conditional trait.

Mailaender committed:
Allow the Saboteur cloak to be manually controlled.

atlimit8 commented on OpenRA.Mods.Common/Traits/Conditions/GrantChargeDrainCondition.cs:
```diff
@@ -0,0 +1,223 @@
+		readonly Actor self;
+		readonly bool canTurn;
+		int deployedToken = ConditionManager.InvalidConditionToken;
+		int deployingToken = ConditionManager.InvalidConditionToken;
```
This is never is never used to grant a condition, but would be revoked below.

atlimit8 commented on OpenRA.Mods.Common/Traits/Conditions/GrantChargeDrainCondition.cs:
```diff
@@ -0,0 +1,223 @@
+				deployedToken = manager.GrantCondition(self, Info.DeployedCondition);
+
+			if (deployingToken != ConditionManager.InvalidConditionToken)
+				deployingToken = manager.RevokeCondition(self, deployingToken);
```
This code will never run since `deployingToken` will always be `ConditionManager.InvalidConditionToken`.

atlimit8 reviewed (changes requested):
I didn't bother testing since the code looks like it is between granting two conditions and granting one condition.

Mailaender commented:
Removed the unnecessary token.

pchote commented on mods/d2k/rules/infantry.yaml:
```diff
@@ -281,11 +281,16 @@ saboteur:
 		CloakDelay: 85
 		CloakSound: STEALTH1.WAV
-		UncloakOn: Attack, Unload, Infiltrate, Demolish, Move, Damage, Heal
+		UncloakOn: Attack, Unload, Infiltrate, Demolish, Damage, Heal
```
How did the original behave when attacking/unloading/infiltrating/damaged/healed while cloaked?

Mailaender replied:
I don't know, but I cleaned this up nonetheless.

pchote replied:
The behaviour on damage and demolishing still remain to be resolved!

Mailaender replied:
@penev92 wanted a reminder.

penev92 replied:
They are now properly set up 👍

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantChargeDrainCondition.cs:
```diff
@@ -0,0 +1,216 @@
+		IResolveOrder, IIssueOrder, ISelectionBar, IOrderVoice, ISync, ITick, IIssueDeployOrder
+	{
+		readonly Actor self;
+		readonly bool canTurn;
```
This doesn't appear to be used for anything?

Mailaender replied:
Removed it.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantChargeDrainCondition.cs:
```diff
@@ -0,0 +1,216 @@
+			if (deployedToken == ConditionManager.InvalidConditionToken)
+				deployedToken = manager.GrantCondition(self, Info.DeployedCondition);
+
+			var drainPercentage = 100 - (ticks * 100 / Info.ChargeTicks);
```
This looks like it may suffer serious precision limitations. Can anyone confirm?

Mailaender replied:
I dumped values and checked that it gives meaningful numbers. That is also why the `* 100` is in front of `ticks`. Wanted this to keep integer math.

Mailaender committed:
Remove conditions which don't make sense.

They have no weapon to attack with,
the game has no transports to unload from
and there are no healers available.

penev92 commented on OpenRA.Mods.Common/Traits/Conditions/GrantChargeDrainCondition.cs:
```diff
@@ -0,0 +1,215 @@
+		public readonly int ChargeTicks = 1000;
+
+		[Desc("The deployed state's length in ticks.")]
+		public readonly int DrainTicks = 100;
```
Can we please refrain from magic values on these two properties?

penev92 commented on OpenRA.Mods.Common/Traits/Conditions/GrantChargeDrainCondition.cs:
```diff
@@ -0,0 +1,215 @@
+
+		public readonly bool ShowSelectionBar = true;
+		public readonly Color ChargingColor = Color.DarkRed;
+		public readonly Color DischargingColor = Color.DarkMagenta;
```
The color make it incredibly hard to see. If we're not going to use pips then I say we stick to the blue used by production structures. (Orange also is a good color for this.)

penev92 reviewed (changes requested):
The `UncloakOn` conditions are now properly set up. 👍 

There is a "Unit lost" notification that plays when the Saboteur demolishes a building, which shouldn't be there. It sounds weird.

There needs to be a minimum threshold for being able to activate the new trait. For D2k saboteurs that would be 100% charged (but that could change with a potential balance pass to make them more potent).

You seem to have made the math here harder than it needs to be. If instead of total charge time and total discharge time (which get really complicated when we can enable/disable at will) we give the trait info a `MaxChargePoints`, `TicksPerChargePoint` and `TicksPerDischargePoint`, then we only need to keep track of two things - current point count and ticks since we last added/subtracted a point.
Then any calculations regarding the readiness to activate or the state of the charge bar come down to `currentPoints/Info.MaxChargePoints`.

penev92 commented:
*(This is more of documentation for future PRs than a change request for this one, though! It is a very serious scope creep that shouldn't happen here.)*
Also saboteurs should uncloak (intended or not, that's what the original does, so I should at least mention it) while entering a building for demolition. But that is currently not supported, so is a part of a bigger rework and is out of scope for this PR. (`Cloak.UncloakFor` doesn't seem to really work for Infiltration and Demolition.
Also they are apparently not targetable when then uncloak to enter, while in ORA uncloaking can get them killed, even while entering a structure, which might pose a noticeable balancing change for the worse.
The problem with making them untargetable while entering in ORA is that they take some between "starting to enter" and actually "entering"/disappearing, while in the original D2k it seems to happen much quicker, allowing for the visual effect of seeing the saboteur, but not having time to do anything about it, thus justifying the non-targetability.
So in Dune it works, but in ORA it is "having the bonus of having them be visible for a second, but purely visual" vs "keeping them invisible to ensure they can get the job done".

abcdefg30 commented:
Are they just untargetable or unkillable as well in the original?

penev92 commented:
This is harder to test in the original than it is in OpenRA, but the AI doesn't even react to them when they start entering a building. And units' autotargetting doesn't kick in. So I suppose unless you force-fire on their location they are virtually unkillable as well? But they are only visible for about 2 seconds, so I'd say effectively both?
Which is what would cause trouble here - unless we lower their "visible" time it would be really weird that you or your units don't and can't react at all.
*(This is more of documentation for future PRs than a change request for this one, though! It is a very serious scope creep that shouldn't happen here.)*

abcdefg30 commented:
Thinking about it, we imho shouldn't make them untargetable/unkillable even if that is what the original did in that case.

Mailaender commented:
I like your simpler charge/drain computation and changed it.

Inq8 commented:
> 
> 
> Also they are apparently not targetable when then uncloak to enter, while in ORA uncloaking can get them killed, even while entering a structure, which might pose a noticeable balancing change for the worse.

Should demolish & infiltrate not just have something like **CaptureManager** that passes the equivalent of **CapturingCondition** to the demolishing/infiltrating actor so you can disable **Targetable** or enable a **DamageMultiplier** with conditions?

penev92 commented:
The entering code and the cloaking code predate conditions, so currently no, but conditions + updating those two are the way to go, yes (not for this PR, ofc - that would be a huge scope creep).

penev92 commented:
At a glance the code still seems to be missing a `MinChargeActivationThreshold` (or whatever you decide to call it) property on `GrantChargeDrainConditionInfo`.

Mailaender commented:
Rebased (not sure why) and added `MinChargeActivationThreshold`.

penev92 commented:
- The CloakDelay should be much shorter. Refer to my 3rd comment in #18051, where I mention Saboteurs and cloaking actors.
 - The max energy default is 1000 but the min activation default is 100. That seems inconsistent.

Other than that, the YAML part looks nicely set up now. I will try to go over the code tomorrow.

Mailaender commented:
What is the charge time exactly? Didn't we already calculate this meticulously? #18051 only lists build times and the support power time is probably also just the spawn time for a new saboteur.

penev92 commented:
I seem to have somehow managed to *NOT* update my 3rd comment on the issue with the second table related to misc charges like saboteur charge/discharge/recloak :(
I will update it, but in the mean time - recloaking should take 25 instead of 85 ticks (it is 2 seconds in the original game, and we more or less agreed on #18051 that we're going for double speed in ORA). The charge and discharge times are as they should be.

Updated https://github.com/OpenRA/OpenRA/issues/18051#issuecomment-629677523 with the second table to the best of my recollection of what I wanted to add there last night.

Mailaender commented:
Rebased this due to https://github.com/OpenRA/OpenRA/pull/17959 and updated the recloak time.

pchote commented:
Ping @penev92 for a rereview.

abcdefg30 commented:
Ping @MustaphaTR and again @penev92 for review.

penev92 commented on OpenRA.Mods.Common/Traits/Conditions/GrantChargeDrainCondition.cs:
```diff
@@ -0,0 +1,211 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
```
This would probably want an update now.

Mailaender replied:
I hope this is the last time I have to change the year in this file...

penev92 commented on OpenRA.Mods.Common/Traits/Conditions/GrantChargeDrainCondition.cs:
```diff
@@ -0,0 +1,211 @@
+		public readonly string DeployedCondition = null;
+
+		[Desc("Time it takes to fully charge up again.")]
+		public readonly int ChargeTicks = 1000;
```
I still think `ChargeTicks` and `DrainTicks` should default to 0 (or -1?) and have `FieldLoader.Require`.

penev92 commented on mods/d2k/rules/infantry.yaml:
```diff
@@ -281,11 +281,16 @@ saboteur:
-		InitialDelay: 85
+		RequiresCondition: deployed
+		InitialDelay: 0
 		CloakDelay: 85
```
Shouldn't this also be 25 ticks?

penev92 commented on OpenRA.Mods.Common/Traits/Conditions/GrantChargeDrainCondition.cs:
```diff
@@ -0,0 +1,211 @@
+		[Sync]
+		int ticks;
+
+		TimedDeployState deployState;
```
Should we sync this as well?

penev92 commented on OpenRA.Mods.Common/Traits/Conditions/GrantChargeDrainCondition.cs:
```diff
@@ -0,0 +1,211 @@
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Grants a condition for a certain time when deployed. Can be reversed anytime " +
+		"by undeploying. Leftover charges are taken into account when recharging.")]
```
Informational (I am fine with not using it for the saboteur):
From a modder's point of view you may want to disable cancelling in some cases. A `IsManuallyUndeployable` property would be nice (defaulting to false).
Dune 2000 actually doesn't allow you to cancel the Saboteur's cloak when active.

penev92 reviewed (commented):
The Min/Max activation thresholds are missing.
Your math is still not what I'd go with - it will do the job for the Saboteur, but adding my ChargePoints suggestion would allow for a lot more flexibility and would tie in nicely with the activation thresholds.

Also can you please rebase? That would make testing easier. After that I can show you a diff of what I'm thinking with the ChargePoints if you'd like.

Mailaender commented:
Rebased and updated a bit.

penev92 committed:
Removed UncloakOn conditions that don't make sense in D2k rules

pchote commented:
Does https://github.com/Mailaender/OpenRA/pull/64 count as a review?

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantChargeDrainCondition.cs:
```diff
@@ -0,0 +1,214 @@
+
+		[FieldLoader.Require]
+		[Desc("Time it takes to fully charge up again.")]
+		public readonly int ChargeTicks = 0;
```
I have a couple of comments on these fields that we expose to modders:
* The naming regresses #9810
* The trait is called `GrantChargeDrainCondition` to match `GrantPrerequisiteChargeDrainPower`, but the configuration it exposes is completely different. This makes things unnecessarily confusing.
* We're missing a way for modders to specify that the condition must fully discharge instead of being interruptable at will.

To fix these, can we please replace `ChargeTicks` and `DrainTicks` with:

```csharp
[FieldLoader.Require]
[Desc("The amount of time in ticks it takes to fully recharge.")]
public readonly int ChargeDuration = 0;

[Desc("Rate at which the condition discharges compared to charging.")]
public readonly int DischargeModifier = 100;

[Desc("Allow the condition to be disabled mid-discharge.")]
public readonly int Interruptable = true;
```

Mailaender committed:
Avoid the "Unit lost" notification.

Mailaender commented:
Merged in @penev92's refactor and renamed fields.

pchote commented:
My other point still stands, so i'll repeat it:
* The trait is called `GrantChargeDrainCondition` to match `GrantPrerequisiteChargeDrainPower`, but the configuration it exposes is completely different. This makes things unnecessarily confusing.

To fix these, can we please replace `ChargeTicks` and `DrainTicks` with:

```csharp
[FieldLoader.Require]
[Desc("The amount of time in ticks it takes to fully recharge.")]
public readonly int ChargeDuration = 0;

[Desc("Rate at which the condition discharges compared to charging.")]
public readonly int DischargeModifier = 100;
```

penev92 commented:
Since it was likely me who suggested the current setup here, I'll step in with my two cents: no, IMO that is completely backwards. My goal here was to make a generic enough trait that doesn't need to be constrained to only one use-case.

pchote commented:
I'm not asking for any functionality to be removed, just for it to be exposed in a consistent way to our existing traits. This trait isn't doing anything new, so should not be reinventing the wheel by inventing new terminology.

penev92 commented:
For the protocol (to be clear to others what the status of this PR is, and for our own future reference so we don't forget), I recently had a discussion with @pchote on the matter about implementation details and this will be blocked on him making a decision about how he wants said details to look. With the current focus on TS features, netcode rework and an upcoming release I expect this could take a while to get priority, so leaving this remainder.

PunkPun commented:
I'm thinking it could be fairly easy to add a generic energy trait to implement saboteur logic, the deploying would be managed via conditions and `GrantConditionOnDeploy`. The only issue I see is that I would need to add `UndeployOnCondition` to `GrantConditionOnDeploy`, so that a unit would forcefully undeploy once the charge runs out

penev92 commented:
This has been booked by design concerns for an yeah and a half now. We can either create a Saboteur-specific trait like we have (had?) for the MAD tank or drop the subject until a truly generic solution comes around. In either case there is no point in keeping this open until then.
